dao-smartcontract/test/Staking.t.sol

```diff
@@ -0,0 +1,436 @@
+
+    function setUp() public {
+        vm.label(OWNER, "OWNER");
+        vm.label(USER_1, "USER_1");
+
+        stakingToken = new MockERC20();
+        rewardsToken = new MockERC20();
+
+        staking = new Staking();
+        staking.initialize(address(stakingToken), address(rewardsToken), OWNER);
+
+        rewardsToken.mint(address(staking), REWARD_PER_BLOCK * BLOCKS_AMOUNT);
+
+        vm.prank(OWNER);
+        staking.setRewards(REWARD_PER_BLOCK, STARTING_BLOCK, BLOCKS_AMOUNT);
+
+        stakingToken.mint(USER_1, INITIAL_BALANCE);
+
+        vm.prank(USER_1);
+        stakingToken.approve(address(staking), INITIAL_BALANCE);
+    }
+
+    // initialize
+    function test_StakingInitialization() public view {
+        assertEq(staking.owner(), OWNER);
+        assertEq(address(staking.stakingToken()), address(stakingToken));
+        assertEq(address(staking.rewardsToken()), address(rewardsToken));
+    }
+
+    // setRewards
+    function test_SetRewards_UpdatesRewards() public view {
+        vm.assertEq(staking.rewardPerTokenStored(), 0);
+        vm.assertEq(staking.lastUpdateBlock(), block.number);
+    }
+
+    function test_SetRewards() public {
+        staking = new Staking();
+        staking.initialize(address(stakingToken), address(rewardsToken), OWNER);
+        rewardsToken.mint(address(staking), REWARD_PER_BLOCK * BLOCKS_AMOUNT);
+
+        vm.prank(OWNER);
+        vm.expectEmit(false, false, false, true);
+        emit RewardsSet(REWARD_PER_BLOCK, STARTING_BLOCK, STARTING_BLOCK + BLOCKS_AMOUNT - 1);
+        staking.setRewards(REWARD_PER_BLOCK, STARTING_BLOCK, BLOCKS_AMOUNT);
+
+        assertEq(staking.rewardPerBlock(), REWARD_PER_BLOCK);
+        assertEq(staking.firstBlockWithReward(), STARTING_BLOCK);
+        assertEq(staking.lastBlockWithReward(), STARTING_BLOCK + BLOCKS_AMOUNT - 1);
+        assertEq(staking.rewardTokensLocked(), REWARD_PER_BLOCK * BLOCKS_AMOUNT);
+    }
+
+    function test_SetRewards_RevertsWhen_NotOwner() public {
+        vm.expectRevert(abi.encodeWithSignature("OwnableUnauthorizedAccount(address)", address(this)));
+        staking.setRewards(REWARD_PER_BLOCK, STARTING_BLOCK, BLOCKS_AMOUNT);
+    }
+
+    function test_SetRewards_FailsWhenNoEnoughRewards() public {
+        staking = new Staking();
+        staking.initialize(address(stakingToken), address(rewardsToken), OWNER);
+
+        vm.prank(OWNER);
+        vm.expectRevert("Not enough tokens for the rewards");
+        staking.setRewards(REWARD_PER_BLOCK, STARTING_BLOCK, BLOCKS_AMOUNT);
+    }
+
+    function test_SetRewards_FailsWhenNoEnoughStakedRewards() public {
+        staking = new Staking();
+        staking.initialize(address(stakingToken), address(stakingToken), OWNER);
+
+        vm.prank(OWNER);
+        vm.expectRevert("Not enough tokens for the rewards");
+        staking.setRewards(REWARD_PER_BLOCK, STARTING_BLOCK, BLOCKS_AMOUNT);
+    }
+
+    // stake
+    function test_Stake() public {
+        vm.prank(USER_1);
```

Comment:
needs to add more users cases

Reply:
Can you give me an example?

Reply:
- Single User Staking Scenario
- Two Users Staking Scenario
- Partial Withdrawal Scenario
- Multiple Stakes and Withdrawals

Reply:
oh I see what you mean now. Thanks!